### ISSUE TITLE
Automatically clean-up Grease Pencil strokes from rough sketches

### ISSUE THREAD

**Issue by b4zz4** [chsh2/nijiGPen#4]
I don't know why I think these add-ons can also solve this.

https://blender.community/c/rightclickselect/WQyE/

A simpler tool could be made. Take two or more lines and join them by averaging their closest points: associating proximity.
When the points exceed a threshold they are "joined"
Creating a new line and deleting the old ones.

![](https://cdn.masto.host/mographsocial/media_attachments/files/109/218/203/715/850/201/small/b95c64d42ee46797.png)

https://cdn.masto.host/mographsocial/media_attachments/files/109/218/192/820/913/542/original/203165e8d2700de4.mp4

This can be done sequentially and in iterations so that the result is more organic and mimics the results of neural networks.

Fast example:

https://cdn.masto.host/mographsocial/media_attachments/files/109/219/323/481/838/949/original/73ce9140b1ac3c42.mp4

(Quickly modifying the magnet stroke)

**Comment by chsh2** [comment]
Yes the sketch cleanup is also a function that I want to implement.

I read the StrokeAggregator paper and tried its program before. However, it was quite slow and took ~10min to process a single drawing in my computer. I do not know if I misconfigured something or the time complexity is indeed high. Maybe we can start from a simpler version of clustering algorithm.

I am also investigating machine learning methods. Though I am not an expert of it, the current popular models such as Transformer look like a good fit to process vector strokes. Now I am generating a dataset, by adding noise to GP strokes and letting the neural network to reverse it. I have not got enough results to show though.

**Comment by b4zz4** [comment]
Only firstly chosen lines could be joined similar to shapes.
Grouping is a really complex problem. For that, you can use the skicit classification.

**Comment by chsh2** [comment]
I have to make several fixes to current functions before developing new ones, so I am not sure when I can start working on this issue. But I did some experiments and would like to share some results here.

- It seems not enough to simply average the closest points. There are many cases where this method does not work, just as stated in the StrokeAggregator paper.
- I tried to train a Transformer network but it also did not work. Maybe I did it in a wrong way, but I cannot come up with a fix for now.
- Then, I tried to implement a simpler version of StrokeAggregator, in combination with some methods from another paper:
  - Lee, In-Kwon. "Curve reconstruction from unorganized points." Computer aided geometric design 17, no. 2 (2000): 161-177.

The last approach works in some cases like the following ones:
![emst](https://user-images.githubusercontent.com/110356534/204160879-0f4f7a44-79ab-4e1a-b028-310a622e34b6.gif)

I think it is a good sign and I will continue. But it is not finished since there are also some failure cases.

**Comment by tugok** [comment]
yooo thats me who suggested on blender community. So nice to see you actually implementing this! Cant wait!

**Comment by chsh2** [comment]
I will release this function in several days, and it is also in the current source code if you want to test it now.

Currently, strokes need to be selected manually. The next step is the implementation of a clustering algorithm. I reread the StrokeAggregator paper and found its clustering algorithm having a very high time complexity. On the other hand, most common clustering algorithms will not work well on curves if simply calculating the pairwise point distance. Therefore, I tend to develop a new method.

**Comment by b4zz4** [comment]
Work perfect:

https://cdn.masto.host/mographsocial/media_attachments/files/109/598/135/688/607/381/original/1f4650f3d4c8ce83.mp4

**Comment by chsh2** [comment]
> Work perfect:
> 


Thank you for your test case. Here one another possible issue is the difference between curves and straight lines. The algorithm tends to create rounded corners, therefore does not work well on straight lines with sharp corners, unless we set a small smoothness value manually. This issue is not mentioned in the research papers. Maybe some additional optimization methods are needed.

**Comment by b4zz4** [comment]
Maybe you could use a softened GS, but it has some problems at the beginning and the end

https://en.wikipedia.org/wiki/Savitzky%E2%80%93Golay_filter
https://scipy-cookbook.readthedocs.io/items/SavitzkyGolay.html

**Comment by b4zz4** [comment]
Will there be a way to detect that points are the originals and take the stroke pressure?. To imitate the original pressure

**Comment by chsh2** [comment]
https://user-images.githubusercontent.com/110356534/215008048-5657f004-f97b-462e-9067-6c574f4f6796.mp4

Made a first implementation of identifying multiple strokes. Maybe some post-processing methods are needed, such as pinching the ends of strokes. Also, I have no confidence for it to work perfectly when the drawing is very complicated. I am thinking of adding another interactive way in the Draw mode.

> Will there be a way to detect that points are the originals and take the stroke pressure?. To imitate the original pressure

Yes that should be possible. I will try to add some options inheriting the point attributes after finishing the main function.

**Comment by b4zz4** [comment]
https://github.com/squidrice21/StripMaker

**Comment by chsh2** [comment]
> https://github.com/squidrice21/StripMaker

Thank you for the information. It looks interesting and I will see if any techniques can be used here. Hopefully I can get more time to work on this project the next month...
